libmcount/mcount.c

```diff
@@ -1762,6 +1762,83 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
 	strv_free(&info.cmds);
 }
 
+/**
+ * deep_copy_filter - perform a deep of a filter structure
+ * @old    - structure to copy
+ * @return - dynamically allocated deep copy
+ */
+static struct uftrace_filter *deep_copy_filter(struct uftrace_filter *old)
+{
+	struct uftrace_filter *new;
+	struct uftrace_arg_spec *arg, *arg_copy;
+
+	/* copy the whole structure */
+	new = xmalloc(sizeof(*new));
+	memcpy(new, old, sizeof(*old));
+
+	/* deep copy nested argspec list */
+	INIT_LIST_HEAD(&new->args);
+	list_for_each_entry(arg, &old->args, list) {
+		arg_copy = xmalloc(sizeof(*arg_copy));
+		memcpy(arg_copy, arg, sizeof(*arg));
+		INIT_LIST_HEAD(&arg_copy->list);
+		if (arg->type_name)
+			arg_copy->type_name = xstrdup(arg->type_name);
+		list_add_tail(&arg_copy->list, &new->args);
+	}
+
+	/* deep copy nested trigger.pargs */
+	new->trigger.pargs = &new->args;
+
+	return new;
+}
+
+/**
+ * __deep_copy_triggers - recursively perform a deep copy of a rbtree with
+ * 'struct uftrace_filter' nodes
+ * @dest - pointer to the address of the copy (modified in place)
+ * @src  - original rbtree
+ */
+static void __deep_copy_triggers(struct rb_node **dest, struct rb_node *src)
+{
+	struct uftrace_filter *old, *new;
+
+	if (!src) {
+		*dest = NULL;
+		return;
+	}
+
+	old = rb_entry(src, struct uftrace_filter, node);
+	new = deep_copy_filter(old);
+	*dest = &new->node;
+
+	if (src->rb_left) {
+		__deep_copy_triggers(&(*dest)->rb_left, src->rb_left);
+		(*dest)->rb_left->rb_parent_color = (unsigned long)new & ~1;
+		(*dest)->rb_left->rb_parent_color |= rb_color(src->rb_left);
+	}
+	if (src->rb_right) {
+		__deep_copy_triggers(&(*dest)->rb_right, src->rb_right);
+		(*dest)->rb_right->rb_parent_color = (unsigned long)new & ~1;
+		(*dest)->rb_right->rb_parent_color |= rb_color(src->rb_right);
+	}
+}
+
+/**
+ * deep_copy_triggers - allocate a new rbtree and store a deep copy of the
+ * original one
+ * @src    - rbtree to copy
+ * @return - deep copy
+ */
+static struct rb_root *deep_copy_triggers(struct rb_root *src)
+{
+	struct rb_root *new;
+	new = xmalloc(sizeof(*new));
+	*new = RB_ROOT;
+	__deep_copy_triggers(&new->rb_node, src->rb_node);
+	return new;
+}
```

Comment:
Can you please move this logic to utils/filter.c?  I think it's better to keep the related code together for the future changes don't miss to update any fields.  Also please add a unit test for the deep copy to verify there's no leak (we run the test with asan).

Reply:
I added a unit test, with a simple value check on the deep copy. I don't check is the tree is valid though.

Reply:
I think that's fine, thanks.